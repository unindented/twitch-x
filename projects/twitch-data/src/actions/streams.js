import {getAction} from '../api'

export function loadTopStreams ({limit}) {
  const action = {
    type: 'LOAD_TOP_STREAMS'
  }
  return getAction(action, '/streams', {query: {limit}})
}
