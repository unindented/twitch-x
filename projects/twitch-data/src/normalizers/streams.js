import {normalize, schema} from 'normalizr'
import {valueOrNull} from './utils'

const stream = new schema.Entity('streams', {}, {
  idAttribute: '_id',
  processStrategy: (stream) => {
    return {
      id: valueOrNull(stream._id),
      status: valueOrNull(stream.channel.status),
      channel: valueOrNull(stream.channel.display_name),
      game: valueOrNull(stream.game),
      image: valueOrNull(stream.preview.template),
      viewers: valueOrNull(stream.viewers)
    }
  }
})

const root = [stream]

export function normalizeTopStreams (payload) {
  return normalize(payload.streams, root)
}

export {normalizeTopStreams as normalizeSearchStreams}
