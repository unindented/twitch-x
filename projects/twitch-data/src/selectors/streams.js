import {createSelector} from 'reselect'

export const getTopStreams = createSelector(
  (state) => state.data.streams.top,
  (state) => state.data.streams.byId,
  (topStreamIds, streamsById) => topStreamIds.map((id) => streamsById[id])
)

export const getTopStreamsForGame = createSelector(
  (state, props) => {
    const game = state.data.games.byId[props.id]
    return (game && game.topStreams) || []
  },
  (state) => state.data.streams.byId,
  (gameStreamIds, streamsById) => gameStreamIds.map((id) => streamsById[id])
)

export const getTopStreamsForCommunity = createSelector(
  (state, props) => {
    const community = state.data.communities.byId[props.id]
    return (community && community.topStreams) || []
  },
  (state) => state.data.streams.byId,
  (communityStreamIds, streamsById) => communityStreamIds.map((id) => streamsById[id])
)

export const getSearchStreams = createSelector(
  (state) => state.data.streams.search,
  (state) => state.data.streams.byId,
  (searchStreamIds, streamsById) => searchStreamIds.map((id) => streamsById[id])
)
