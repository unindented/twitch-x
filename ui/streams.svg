<?xml version="1.0" encoding="UTF-8"?>
<svg width="256px" height="256px" viewBox="0 0 256 256" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <defs></defs>
    <g id="Assets" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <g id="Streams" fill-rule="nonzero" fill="#FFFFFF">
            <path d="M181.333333,112 L181.333333,74.6666667 C181.333333,68.8 176.533333,64 170.666667,64 L42.6666667,64 C36.8,64 32,68.8 32,74.6666667 L32,181.333333 C32,187.2 36.8,192 42.6666667,192 L170.666667,192 C176.533333,192 181.333333,187.2 181.333333,181.333333 L181.333333,144 L224,186.666667 L224,69.3333333 L181.333333,112 Z" id="Shape"></path>
        </g>
    </g>
</svg>