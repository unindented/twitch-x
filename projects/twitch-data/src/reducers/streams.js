import update from 'immutability-helper'
import {normalizeTopStreams, normalizeSearchStreams} from '../normalizers/streams'

const initialState = {
  byId: {},
  top: [],
  search: []
}

export default (state = initialState, action = {}) => {
  switch (action.type) {
    case 'LOAD_TOP_STREAMS_SUCCESS': {
      const {entities, result} = normalizeTopStreams(action.payload.response)
      return update(state, {
        byId: {$merge: entities.streams},
        top: {$set: result}
      })
    }
    case 'LOAD_SEARCH_STREAMS_SUCCESS': {
      const {entities, result} = normalizeSearchStreams(action.payload.response)
      return update(state, {
        byId: {$merge: entities.streams},
        search: {$set: result}
      })
    }
  }

  return state
}
