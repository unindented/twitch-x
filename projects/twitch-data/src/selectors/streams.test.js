import {getTopStreams, getSearchStreams} from './streams'

describe('selectors/streams', () => {
  describe('.getTopStreams', () => {
    it('returns the top streams when available', () => {
      const state = {
        data: {
          streams: {
            byId: {
              1: {channel: 'foo'},
              2: {channel: 'bar'},
              3: {channel: 'baz'}
            },
            top: [1, 2]
          }
        }
      }
      expect(getTopStreams(state)).toMatchSnapshot()
    })
  })

  describe('.getSearchStreams', () => {
    it('returns the search streams when available', () => {
      const state = {
        data: {
          streams: {
            byId: {
              1: {channel: 'foo'},
              2: {channel: 'bar'},
              3: {channel: 'baz'}
            },
            search: [1, 2]
          }
        }
      }
      expect(getSearchStreams(state)).toMatchSnapshot()
    })
  })
})
