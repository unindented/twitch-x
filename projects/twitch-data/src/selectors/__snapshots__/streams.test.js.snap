// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors/streams .getSearchStreams returns the search streams when available 1`] = `
Array [
  Object {
    "channel": "foo",
  },
  Object {
    "channel": "bar",
  },
]
`;

exports[`selectors/streams .getTopStreams returns the top streams when available 1`] = `
Array [
  Object {
    "channel": "foo",
  },
  Object {
    "channel": "bar",
  },
]
`;

exports[`selectors/streams .getTopStreamsForCommunity returns an empty array when the community cannot be found 1`] = `Array []`;

exports[`selectors/streams .getTopStreamsForCommunity returns the top streams when available 1`] = `
Array [
  Object {
    "channel": "foo",
  },
  Object {
    "channel": "bar",
  },
]
`;

exports[`selectors/streams .getTopStreamsForGame returns an empty array when the game cannot be found 1`] = `Array []`;

exports[`selectors/streams .getTopStreamsForGame returns the top streams when available 1`] = `
Array [
  Object {
    "channel": "foo",
  },
  Object {
    "channel": "bar",
  },
]
`;
